File: lib/factorix/types/info_json.rb
# frozen_string_literal: true

require "json"
require "tempfile"
require "zip"

module Factorix
  module Types
    InfoJSON = Data.define(:name, :version, :title, :author, :description, :factorio_version, :dependencies)

    # Factorio mod info.json representation
    #
    # Represents the metadata file that must be present in every Factorio mod.
    # Only required fields (name, version, title, author) are enforced.
    #
    # @see https://lua-api.factorio.com/latest/auxiliary/mod-structure.html
    class InfoJSON
      # Parse info.json from JSON string
      #
      # @param json_string [String] JSON content
      # @return [InfoJSON] parsed info.json
      # @raise [ArgumentError] if required fields are missing or JSON is invalid
      def self.from_json(json_string)
        data = JSON.parse(json_string)

        # Validate required fields
        required_fields = %w[name version title author]
        missing = required_fields - data.keys
        raise ArgumentError, "Missing required fields: #{missing.join(", ")}" unless missing.empty?

        # Parse dependencies
        parser = Dependency::Parser.new
        dependencies = (data["dependencies"] || []).map {|dep_str| parser.parse(dep_str) }

        new(name: data["name"], version: MODVersion.from_string(data["version"]), title: data["title"], author: data["author"], description: data["description"] || "", factorio_version: data["factorio_version"], dependencies:)
      rescue JSON::ParserError => e
        raise ArgumentError, "Invalid JSON: #{e.message}"
      end

      # Extract from zip file
      #
      # Uses caching to avoid repeated ZIP extraction for the same file.
      # Cache key is based on file path (MOD ZIPs are immutable after download).
      #
      # @param zip_path [Pathname] path to mod zip file
      # @return [InfoJSON] parsed info.json from zip
      # @raise [ArgumentError] if zip is invalid or info.json not found
      def self.from_zip(zip_path)
        cache = Application.resolve(:info_json_cache)
        logger = Application.resolve(:logger)
        cache_key = cache.key_for(zip_path.to_s)

        # Try to read from cache
        if (cached_json = cache.read(cache_key, encoding: Encoding::UTF_8))
          logger.debug("info.json cache hit", path: zip_path.to_s)
          return from_json(cached_json)
        end

        logger.debug("info.json cache miss", path: zip_path.to_s)

        # Extract from ZIP
        json_string = Zip::File.open(zip_path) {|zip_file|
          info_entry = zip_file.find {|entry| entry.name.end_with?("/info.json") }
          raise ArgumentError, "info.json not found in #{zip_path}" unless info_entry

          info_entry.get_input_stream.read
        }

        # Store in cache
        temp_file = Tempfile.new("info_json_cache")
        begin
          temp_file.write(json_string)
          temp_file.close
          cache.store(cache_key, Pathname(temp_file.path))
          logger.debug("Stored info.json in cache", path: zip_path.to_s)
        ensure
          temp_file.unlink
        end

        from_json(json_string)
      rescue Zip::Error => e
        raise ArgumentError, "Invalid zip file: #{e.message}"
      end
    end
  end
end

File: internal/mod/info_json.go
package mod

import (
	"archive/zip"
	"encoding/json"
	"io"
	"strings"
)

// InfoJSON is the metadata file present in every Factorio MOD.
// Only the required fields (name, version, title, author) are enforced.
//
// See https://lua-api.factorio.com/latest/auxiliary/mod-structure.html
type InfoJSON struct {
	Name            string
	Version         MODVersion
	Title           string
	Author          string
	Description     string
	FactorioVersion string
	Dependencies    []string // raw dependency strings; parsing arrives with internal/dependency
}

// Pointer fields distinguish absent required fields from empty strings.
type infoJSONRaw struct {
	Name            *string  `json:"name"`
	Version         *string  `json:"version"`
	Title           *string  `json:"title"`
	Author          *string  `json:"author"`
	Description     string   `json:"description"`
	FactorioVersion string   `json:"factorio_version"`
	Dependencies    []string `json:"dependencies"`
}

// ParseInfoJSON parses info.json content.
func ParseInfoJSON(data []byte) (InfoJSON, error) {
	var raw infoJSONRaw
	if err := json.Unmarshal(data, &raw); err != nil {
		return InfoJSON{}, &FileFormatError{Msg: "invalid JSON: " + err.Error()}
	}

	var missing []string
	for _, field := range []struct {
		name  string
		value *string
	}{
		{"name", raw.Name},
		{"version", raw.Version},
		{"title", raw.Title},
		{"author", raw.Author},
	} {
		if field.value == nil {
			missing = append(missing, field.name)
		}
	}
	if len(missing) > 0 {
		return InfoJSON{}, &FileFormatError{Msg: "missing required fields: " + strings.Join(missing, ", ")}
	}

	version, err := ParseMODVersion(*raw.Version)
	if err != nil {
		return InfoJSON{}, &FileFormatError{Msg: err.Error()}
	}

	return InfoJSON{
		Name:            *raw.Name,
		Version:         version,
		Title:           *raw.Title,
		Author:          *raw.Author,
		Description:     raw.Description,
		FactorioVersion: raw.FactorioVersion,
		Dependencies:    raw.Dependencies,
	}, nil
}

// InfoJSONFromZIP extracts and parses info.json from a MOD ZIP file.
func InfoJSONFromZIP(path string) (InfoJSON, error) {
	zr, err := zip.OpenReader(path)
	if err != nil {
		return InfoJSON{}, &FileFormatError{Path: path, Msg: "invalid zip file: " + err.Error()}
	}
	defer zr.Close()

	// MOD ZIPs contain a single top-level directory, so info.json always
	// sits below it — hence the "/info.json" suffix match.
	for _, f := range zr.File {
		if !strings.HasSuffix(f.Name, "/info.json") {
			continue
		}
		rc, err := f.Open()
		if err != nil {
			return InfoJSON{}, &FileFormatError{Path: path, Msg: err.Error()}
		}
		data, err := io.ReadAll(rc)
		rc.Close()
		if err != nil {
			return InfoJSON{}, &FileFormatError{Path: path, Msg: err.Error()}
		}
		return ParseInfoJSON(data)
	}
	return InfoJSON{}, &FileFormatError{Path: path, Msg: "info.json not found"}
}
